File: Дискретная математика/РГР/Волейбольная сетка_В-18.py
# Задание №18
# Дана волейбольная сетка с заданным количеством ячеек. 
# Определить максимальное число веревок, составляющих ее ребра, которые можно разрезать так, чтобы она не распалась.

print("Введите размеры сетки через пробел (два числа): ")
n, m = [int(el) for el in input().split()]
vertexes = (n + 1) * (m + 1)

matrix = [[0 for _ in range(vertexes)] for _ in range(vertexes)]

# Создание матрицы смежности (не симметрична, только верхний треугольник) и поиск ребер
edges = []
for i in range(vertexes):
    for j in range(i + 1, vertexes):
        if (j - i == 1 and (i + 1) % (m + 1) != 0) or (j - i == m + 1):
            matrix[i][j] = 1
            edges.append([i, j])

# Поиск остова
components_of_vertexes = [x for x in range(vertexes)]
skeleton = []
for start, end in edges:
    if components_of_vertexes[start] != components_of_vertexes[end]:
        skeleton.append([start, end])
        max_comp = max(components_of_vertexes[start], components_of_vertexes[end]) # Наибольший номер компоненты связности
        min_comp = min(components_of_vertexes[start], components_of_vertexes[end]) # Наименьший номер компоненты связности
        for i in range(len(components_of_vertexes)):
            if components_of_vertexes[i] == max_comp:
                components_of_vertexes[i] = min_comp

result = len(edges) - len(skeleton)
print(f"Максимальное число веревок, которые можно разрезать так, чтобы она не распалась: {result}")

File: Алгоритмизация и программирование/2 семестр/РГР/Волейбольная сетка_В-18.cs
//Задание №18
//Дана волейбольная сетка с заданным количеством ячеек. 
//Определить максимальное число веревок, составляющих ее ребра, которые можно разрезать так, чтобы она не распалась.

using System;

namespace VolleyballNet
{
    class Program
    {
        static void Main()
        {
            Console.Write("Введите размеры сетки через пробел (два числа): ");
            string[] data = Console.ReadLine().Split();
            int n = Convert.ToInt32(data[0]);
            int m = Convert.ToInt32(data[1]);
            int vertexes = (n + 1) * (m + 1);

            int[,] matrix = new int[vertexes, vertexes];

            List<Tuple<int, int>> edges = new List<Tuple<int, int>>();
            for (int i = 0; i < vertexes; i++)
            {
                for (int j = i + 1; j < vertexes; j++)
                {
                    if ((j - i == 1 && (i + 1) % (m + 1) != 0) || (j - i == m + 1))
                    {
                        matrix[i, j] = 1;
                        edges.Add(Tuple.Create(i, j));
                    }
                }
            }

            int[] components = new int[vertexes];
            for (int i = 0; i < vertexes; i++)
            {
                components[i] = i;
            }

            List<Tuple<int, int>> skeleton = new List<Tuple<int, int>>();

            foreach (var edge in edges)
            {
                int start = edge.Item1;
                int end = edge.Item2;

                if (components[start] != components[end])
                {
                    skeleton.Add(edge);
                    int max_comp = Math.Max(components[start], components[end]);
                    int min_comp = Math.Min(components[start], components[end]);

                    for (int i = 0; i < vertexes; i++)
                    {
                        if (components[i] == max_comp)
                        {
                            components[i] = min_comp;
                        }
                    }
                }
            }

            int result = edges.Count - skeleton.Count;
            Console.WriteLine($"Максимальное число веревок, которые можно разрезать так, чтобы она не распалась: {result}");
        }
    }
}
